pkg/network/go/bininspect/symbols.go

```diff
@@ -0,0 +1,297 @@
+	minSymbolNameSize, maxSymbolNameSize := getSymbolBoundaries(copyOfWantedSymbols)
+	// Pre-allocating a buffer to read the symbol string into.
+	// The size of the buffer is maxSymbolNameSize + 1, for null termination.
+	symbolNameBuf := make([]byte, maxSymbolNameSize+1)
+	// Pre allocating a buffer for reading the symbol entry in the string table.
+	allocatedBufferForSymbolNameRead := make([]byte, 4)
+	// Pre allocating a buffer for reading the rest of the symbol fields from the symbol section.
+	allocatedBufferForSymbolRead := make([]byte, elf.Sym64Size-len(allocatedBufferForSymbolNameRead))
+
+	// Iterating through the symbol table. We skip the first elf.Sym64Size bytes as they are zeros.
+	for readLocation := int64(elf.Sym64Size); uint64(readLocation) < symbolSection.Size; readLocation += elf.Sym64Size {
+		// Reading the symbol entry in the string table.
+		stringEntry, ok := readSymbolEntryInStringTable(symbolSection.ReaderAt, f.ByteOrder, readLocation, allocatedBufferForSymbolNameRead)
+		if !ok {
+			continue
+		}
+
+		// Trying to get string representation of symbol.
+		// If the symbol name's length is not in the boundaries [minSymbolNameSize, maxSymbolNameSize+1] then we fail,
+		// and continue to the next symbol.
+		symbolName, ok := getSymbolNameByEntry(f.Sections[symbolSection.Link].ReaderAt, stringEntry, minSymbolNameSize, symbolNameBuf)
+		if !ok {
+			continue
+		}
+
+		// Checking the symbol is relevant for us.
+		if _, ok := copyOfWantedSymbols[symbolName]; !ok {
+			continue
+		}
+
+		// Complete the symbol reading.
+		var symbol elf.Symbol
+		if err := readRestOfSymbol64(&symbol, symbolSection.ReaderAt, f.ByteOrder, symbolName, readLocation+4, allocatedBufferForSymbolRead); err != nil {
+			continue
+		}
+		symbols = append(symbols, symbol)
+		// If relevant, delete it for optimization purposes.
+		delete(copyOfWantedSymbols, symbolName)
+
+		// If no symbols left, stop running.
+		if len(copyOfWantedSymbols) == 0 {
+			break
+		}
+	}
+
+	return symbols, nil
+}
+
+// getSymbols32 extracts the given symbol list from the binary.
+func getSymbols32(f *elf.File, typ elf.SectionType, wantedSymbols common.StringSet) ([]elf.Symbol, error) {
+	// Getting the relevant symbol section.
+	symbolSection := f.SectionByType(typ)
+	if symbolSection == nil {
+		return nil, elf.ErrNoSymbols
+	}
+
+	// Checking the symbol section size is aligned to a multiplication of Sym64Size.
+	if symbolSection.Size%elf.Sym32Size != 0 {
+		return nil, errors.New("length of symbol section is not a multiple of Sym32Size")
+	}
+
+	// Checking the symbol section link is valid.
+	if symbolSection.Link <= 0 || symbolSection.Link >= uint32(len(f.Sections)) {
+		return nil, errors.New("section has invalid string table link")
+	}
+
+	// Allocating entries for all wanted symbols.
+	symbols := make([]elf.Symbol, 0, len(wantedSymbols))
+
+	// Copying the wanted symbol set, so we can modify it during runtime.
+	copyOfWantedSymbols := wantedSymbols.Clone()
```

Comment:
add more details why do you need the clone of the set.
my understand it is done for optimization purposes to remove the symbols from the set and reduce further iterations, if that's the case I think you can remove this code completely as it just complicates the function and I don't think it adds much perfomance gain, since it is not a large set.

Reply:
removed